### ISSUE TITLE
Make the left menu configurable (in Curses and WebUI)

### ISSUE THREAD

**Issue by nicolargo** [nicolargo/glances#2648]
For the moment, the list of plugins displayed in the UI is "hard coded":

- glances/outputs/glances_curses.py (in the _left_sidebar variable)
- glances/outputs/static/js/App.vue (in the HTML code)

The goal of this PR is to make this list configurable from the configuration file (in the outputs section).

**Comment by nicolargo** [comment]
Look pretty simple for the Curses interface, but do not know how to do it for the WebUI... @notFloran Any idea/tips ?

**Comment by nicolargo** [comment]
Copilot helps:

nicolargo: Please refactor this code

GitHub Copilot: Sure, what would you like me to do?

nicolargo: Refactor the code  by reducing the complexity

GitHub Copilot: One way to reduce the complexity of this code is to extract the actions performed by each key into separate functions. Here's an example of how this could be done:

```python
def __catch_key(self, return_to_browser=False):
    # Catch the pressed key
    self.pressedkey = self.get_key(self.term_window)
    if self.pressedkey == -1:
        return -1

    # Actions (available in the global hotkey dict)...
    logger.debug("Keypressed (code: {})".format(self.pressedkey))
    for hotkey in self._hotkeys:
        if self.pressedkey == ord(hotkey) and 'switch' in self._hotkeys[hotkey]:
            self._handle_switch(hotkey)
        elif self.pressedkey == ord(hotkey) and 'sort_key' in self._hotkeys[hotkey]:
            self._handle_sort_key(hotkey)

    # Other actions...
    if self.pressedkey == ord('\n'):
        self._handle_enter()
    elif self.pressedkey == ord('4'):
        self._handle_quicklook()
    elif self.pressedkey == ord('5'):
        self._handle_top_menu()
    elif self.pressedkey == ord('9'):
        self._handle_theme()
    elif self.pressedkey == ord('e') and not self.args.programs:
        self._handle_process_extended()
    elif self.pressedkey == ord('E'):
        self._handle_erase_filter()
    elif self.pressedkey == ord('f'):
        self._handle_fs_stats()
    elif self.pressedkey == ord('+'):
        self._handle_increase_nice()
    elif self.pressedkey == ord('-'):
        self._handle_decrease_nice()
    elif self.pressedkey == ord('k') and not self.args.disable_cursor:
        self._handle_kill_process()
    elif self.pressedkey == ord('w'):
        self._handle_clean_logs()
    elif self.pressedkey == ord('x'):
        self._handle_clean_critical_logs()
    elif self.pressedkey == ord('z'):
        self._handle_disable_process()
    elif self.pressedkey == curses.KEY_LEFT:
        self._handle_sort_left()
    elif self.pressedkey == curses.KEY_RIGHT:
        self._handle_sort_right()
    elif self.pressedkey == curses.KEY_UP or self.pressedkey == 65 and not self.args.disable_cursor:
        self._handle_cursor_up()
    elif self.pressedkey == curses.KEY_DOWN or self.pressedkey == 66 and not self.args.disable_cursor:
        self._handle_cursor_down()
    elif self.pressedkey == ord('\x1b') or self.pressedkey == ord('q'):
        self._handle_quit(return_to_browser)
    elif self.pressedkey == curses.KEY_F5 or self.pressedkey == 18:
        self._handle_refresh()

    # Return the key code
    return self.pressedkey

def _handle_switch(self, hotkey):
    option = '_'.join(self._hotkeys[hotkey]['switch'].split('_')[1:])
    if self._hotkeys[hotkey]['switch'].startswith('disable_'):
        if getattr(self.args, self._hotkeys[hotkey]['switch']):
            enable(self.args, option)
        else:
            disable(self.args, option)
    elif self._hotkeys[hotkey]['switch'].startswith('enable_'):
        if getattr(self.args, self._hotkeys[hotkey]['switch']):
            disable(self.args, option)
        else:
            enable(self.args, option)
    else:
        setattr(
            self.args,
            self._hotkeys[hotkey]['switch'],
            not getattr(self.args, self._hotkeys[hotkey]['switch']),
        )

def _handle_sort_key(self, hotkey):
    glances_processes.set_sort_key(
        self._hotkeys[hotkey]['sort_key'], self._hotkeys[hotkey]['sort_key'] == 'auto'
    )

def _handle_enter(self):
    self.edit_filter = not self.edit_filter

def _handle_quicklook(self):
    self.args.full_quicklook = not self.args.full_quicklook
    if self.args.full_quicklook:
        self.enable_fullquicklook()
    else:
        self.disable_fullquicklook()

def _handle_top_menu(self):
    self.args.disable_top = not self.args.disable_top
    if self.args.disable_top:
        self.disable_top()
    else:
        self.enable_top()

def _handle_theme(self):
    self._init_colors()

def _handle_process_extended(self):
    self.args.enable_process_extended = not self.args.enable_process_extended
    if not self.args.enable_process_extended:
        glances_processes.disable_extended()
    else:
        glances_processes.enable_extended()
    self.args.disable_cursor = self.args.enable_process_extended and self.args.is_standalone

def _handle_erase_filter(self):
    glances_processes.process_filter = None

def _handle_fs_stats(self):
    self.args.disable_fs = not self.args.disable_fs
    self.args.disable_folders = not self.args.disable_folders

def _handle_increase_nice(self):
    self.increase_nice_process = not self.increase_nice_process

def _handle_decrease_nice(self):
    self.decrease_nice_process = not self.decrease_nice_process

def _handle_kill_process(self):
    self.kill_process = not self.kill_process

def _handle_clean_logs(self):
    glances_events.clean()

def _handle_clean_critical_logs(self):
    glances_events.clean(critical=True)

def _handle_disable_process(self):
    self.args.disable_process = not self.args.disable_process
    if self.args.disable_process:
        glances_processes.disable()
    else:
        glances_processes.enable()

def _handle_sort_left(self):
    next_sort = (self.loop_position() - 1) % len(self._sort_loop)
    glances_processes.set_sort_key(self._sort_loop[next_sort], False)

def _handle_sort_right(self):
    next_sort = (self.loop_position() + 1) % len(self._sort_loop)
    glances_processes.set_sort_key(self._sort_loop[next_sort], False)

def _handle_cursor_up(self):
    if self.args.cursor_position > 0:
        self.args.cursor_position -= 1

def _handle_cursor_down(self):
    if self.args.cursor_position < glances_processes.processes_count:
        self.args.cursor_position += 1

def _handle_quit(self, return_to_browser):
    if return_to_browser:
        logger.info("Stop Glances client and return to the browser")
    else:
        logger.info("Stop Glances (keypressed: {})".format(self.pressedkey))

def _handle_refresh(self):
    pass
```

nicolargo: is it possible to make the components list configurable ?

GitHub Copilot: Yes, it is possible to make the components list configurable. You can create a configuration file (e.g., `config.json`) that contains an array of the components you want to use. Then, you can import this file and dynamically import the components based on the configuration.

Here is an example of how you can do this:

1. Create a `config.json` file:

```json
{
  "components": [
    "help",
    "plugin-alert",
    "plugin-cloud",
    "plugin-connections",
    "plugin-cpu",
    "plugin-diskio",
    "plugin-containers",
    "plugin-folders",
    "plugin-fs",
    "plugin-gpu",
    "plugin-ip",
    "plugin-irq",
    "plugin-load",
    "plugin-mem",
    "plugin-mem-more",
    "plugin-memswap",
    "plugin-network",
    "plugin-now",
    "plugin-percpu",
    "plugin-ports",
    "plugin-process",
    "plugin-quicklook",
    "plugin-raid",
    "plugin-smart",
    "plugin-sensors"
  ]
}
```

2. In your `App.vue` file, import the configuration file and dynamically import the components:

```vue
<script>
import { GlancesStats } from './services.js';
import { store } from './store.js';
import config from './config.json';

export default {
  name: 'App',
  data() {
    return {
      components: {},
    };
  },
  created() {
    this.loadComponents();
  },
  methods: {
    async loadComponents() {
      for (const component of config.components) {
        this.components[component] = () => import(`./components/${component}.vue`);
      }
    },
  },
};
</script>
```

3. Then, you can use the components in your template like this:

```vue
<template>
  <div id="app">
    <component v-for="component in components" :is="component" />
  </div>
</template>
```

Please note that this is a simplified example and you might need to adjust it based on your actual needs and project structure.

nicolargo: Can you add the v-if option in the template ?

GitHub Copilot: Sure, you can add the `v-if` directive to the component in the template. This directive will conditionally render the component based on the truthiness of the expression. Here's how you can do it:

```vue
<template>
  <div id="app">
    <div class="container-fluid">
      <div class="row">
        <div class="col-sm-6 sidebar" v-if="!args.disable_left_sidebar">
          <div class="table">
            <component
              v-for="component in components"
              :is="component"
              :id="`plugin-${component}`"
              class="plugin table-row-group"
              v-if="!args[`disable_${component}`]"
              :data="data"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</template>
```

In this example, the `v-if` directive is used to conditionally render each component based on the corresponding property in the `args` object. The property name is constructed using a template literal: `` `disable_${component}` ``. If the property value is truthy, the component will not be rendered.

**Comment by nicolargo** [comment]
The default plugin list for the left menu is defined:
- glances/outputs/glances_curses.py (in the _left_sidebar variable) for the curses UI
- glances/outputs/static/js/uiconfig.json for the WebUI

The list can be overwrite from the glances.conf file:

```
[outputs]
# Set the the Curses and WebUI interface left menu plugin list (comma-separated)
left_menu=network,wifi,connections,ports,diskio,fs,irq,folders,raid,smart,sensors,now
```

Merged into develop branch.

Will be released in Glances 4.0.0.

**Comment by nicolargo** [comment]
Example:

```
[outputs]
# Set the the Curses and WebUI interface left menu plugin list (comma-separated)
left_menu=network,diskio
``` 
displays:

![image](https://github.com/nicolargo/glances/assets/776747/db67403b-de23-4edb-a503-84c6e8235cf2)
